TypeCobol.Analysis/Cfg/ControlFlowGraphBuilder.cs

```diff
@@ -0,0 +1,2695 @@
+            }
+            if (CurrentDeclarativesContext != null)
+            {//This block is created in the context of a Declaratives.
+                block.SetFlag(BasicBlock<Node, D>.Flags.Declaratives, true);
+            }
+            return block;
+        }
+
+        /// <summary>
+        /// Group Index Counter
+        /// </summary>
+        private int GroupCounter = 0;
+        /// <summary>
+        /// Create a Group Basic Block for a node
+        /// </summary>
+        /// <param name="node">The leading node of the block</param>
+        /// <param name="addToCurrentSentence">true if the block must be added to the current Sentence, false otherwise.</param>
+        /// <returns>The new Block</returns>
+        internal BasicBlockForNodeGroup CreateGroupBlock(Node node, bool addToCurrentSentence)
+        {
+            var block = new BasicBlockForNodeGroup();
+            block.GroupIndex = ++GroupCounter;
+            block.Index = this.CurrentProgramCfgBuilder.Cfg.AllBlocks.Count;
+            this.CurrentProgramCfgBuilder.Cfg.AllBlocks.Add(block);
+            if (node != null)
+            {
+                this.CurrentProgramCfgBuilder.Cfg.BlockFor[node] = block;
+                block.Instructions.AddLast(node);
+            }
+            if (addToCurrentSentence && this.CurrentProgramCfgBuilder.CurrentSentence != null)
+            {
+                this.CurrentProgramCfgBuilder.CurrentSentence.AddBlock(block);
+            }
+            if (CurrentDeclarativesContext != null)
+            {//This group is created in the context of a Declaratives.
+                block.SetFlag(BasicBlock<Node, D>.Flags.Declaratives, true);
+            }
+            return block;
+        }
+
+        /// <summary>
+        /// Create a Fresh Control Flow Graph Builder.
+        /// </summary>
+        /// <returns>The fresh Control Flow Graph Builder</returns>
+        protected virtual ControlFlowGraphBuilder<D> CreateFreshControlFlowGraphBuilder(ControlFlowGraphBuilder<D> parentCfgBuilder = null)
+        {
+            return new ControlFlowGraphBuilder<D>();
+        }
+
+        /// <summary>
+        /// Initialize the Cfg by a Program.
+        /// </summary>
+        /// <param name="program">The target program of the Cfg</param>
+        protected virtual void InitializeCfg(Program program)
+        {
+            Cfg = new ControlFlowGraph<Node, D>();
+            Cfg.ProgramNode = program;
+        }
+
+        /// <summary>
+        /// Initialize the Cfg by a Function.
+        /// </summary>
+        /// <param name="funDecl">The target function declaration of the Cfg</param>
+        protected virtual void InitializeCfg(FunctionDeclaration funDecl)
+        {
+            Cfg = new ControlFlowGraph<Node, D>();
+            Cfg.ProgramNode = funDecl;
+        }
+
+        public static readonly string ROOT_SECTION_NAME = "<< RootSection >>";
```

Comment:
Is this required ? I mean, in Semantic Domain and according to the Cobol spec, when a paragraph is not inside a section, it is attached to the Program directly. This is more problematic for sentences as they still need to be attached to something...

Reply:
Yes, this for a CFG construction point of view, we need to introduce a kind of virtual  ROOT Section, with a start section and a Start Basic Block. See it as Root entities of the graph.